codec-dns/src/main/java/io/netty/handler/codec/dns/AbstractDnsMessage.java

```diff
@@ -0,0 +1,462 @@
+            if (recordOrList instanceof ReferenceCounted) {
+                ((ReferenceCounted) recordOrList).release();
+            } else if (recordOrList instanceof List) {
+                @SuppressWarnings("unchecked")
+                List<DnsRecord> list = (List<DnsRecord>) recordOrList;
+                if (!list.isEmpty()) {
+                    for (Object r : list) {
+                        ReferenceCountUtil.release(r);
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    public DnsMessage touch() {
+        return (DnsMessage) super.touch();
+    }
+
+    @Override
+    public DnsMessage touch(Object hint) {
+        if (leak != null) {
+            leak.record(hint);
+        }
+        return this;
+    }
+
+    @Override
+    public DnsMessage retain() {
+        return (DnsMessage) super.retain();
+    }
+
+    @Override
+    public DnsMessage retain(int increment) {
+        return (DnsMessage) super.retain(increment);
+    }
+
+    @Override
+    protected void deallocate() {
+        clear();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (!(obj instanceof DnsMessage)) {
+            return false;
+        }
+
+        final DnsMessage that = (DnsMessage) obj;
+        if (id() != that.id()) {
+            return false;
+        }
+
+        if (this instanceof DnsQuery) {
+            if (!(that instanceof DnsQuery)) {
+                return false;
+            }
+        } else if (that instanceof DnsQuery) {
+            return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    public int hashCode() {
+        return id() * 31 + (this instanceof DnsQuery? 0 : 1);
+    }
+
+    private Object sectionAt(int section) {
+        switch (section) {
+        case 0:
+            return questions;
+        case 1:
+            return answers;
+        case 2:
+            return authorities;
+        case 3:
+            return additionals;
+        }
+
+        throw new Error(); // Should never reach here.
+    }
+
+    private void setSection(int section, Object value) {
+        switch (section) {
+        case 0:
+            questions = value;
+            return;
+        case 1:
+            answers = value;
+            return;
+        case 2:
+            authorities = value;
+            return;
+        case 3:
+            additionals = value;
+            return;
+        }
+
+        throw new Error(); // Should never reach here.
+    }
+
+    private static int sectionOrdinal(DnsSection section) {
+        return checkNotNull(section, "section").ordinal();
+    }
+
+    private static DnsRecord checkQuestion(int section, DnsRecord record) {
+        if (section == SECTION_QUESTION && !(checkNotNull(record, "record") instanceof DnsQuestion)) {
+            throw new IllegalArgumentException(
+                    "record: " + record + " (expected: " + DnsQuestion.class.getSimpleName() + ')');
```

Comment:
It is raised only when a user application has a bug. i.e. it is never triggered by bad data.